File: akatpp-x/akat_x_gpio_safe_output.py
# Like X_GPIO_OUTPUT$ but times out if state is not updated within state_timeout_deciseconds.
# If timeout happens, then the state of pin is set to 'safe_state'

class Macro:
    def render(self, inv):
        ctx = akat.prepare(inv,
                           required_args = ["object_name", "pin_name"],
                           required_kvs = ["safe_state", "state_timeout_deciseconds"])

        rendered = akat.render(self,
                               object_name = ctx.object_name,
                               pin_name = ctx.pin_name,
                               safe_state = ctx.safe_state,
                               state_timeout_deciseconds = ctx.state_timeout_deciseconds)

        return akat.transform(rendered)

File: akatpp-x/akat_x_gpio_safe_output.c
// Like X_GPIO_OUTPUT$ but times out if state is not updated witjin state_timeout_deciseconds.
// If timeout happens, then the state of pin is set to 'safe_state'

X_GPIO_OUTPUT$(${object_name}__output, ${pin_name})

X_INIT$(${object_name}__init) {
    ${object_name}__output.set(${safe_state});
}

GLOBAL$() {
    STATIC_VAR$(u8 ${object_name}__updated_ago, initial = 255);
}

OBJECT$(${object_name}) {
    METHOD$(void set(u8 state), inline) {
        ${object_name}__output.set(state);
        ${object_name}__updated_ago = 0;
    }
}

X_EVERY_DECISECOND$(${object_name}__ticker) {
    ${object_name}__updated_ago += 1;
    if (!${object_name}__updated_ago) {
        ${object_name}__updated_ago -= 1;
    }
    
    if (${object_name}__updated_ago >= ${state_timeout_deciseconds}) {
        ${object_name}__output.set(${safe_state});
    }
}